File: python/cgm-ai4.py
import numpy as np

def conjugate_gradient(A, b, tol=1e-10, max_iter=1000):
    n = len(b)
    x = np.zeros(n)  # Initial guess
    r = b - np.dot(A, x)
    p = r.copy()
    rsold = np.dot(r, r)

    for i in range(max_iter):
        Ap = np.dot(A, p)
        alpha = rsold / np.dot(p, Ap)
        x += alpha * p
        r -= alpha * Ap
        rsnew = np.dot(r, r)

        if np.sqrt(rsnew) < tol:
            print(f"Convergence achieved after {i+1} iterations.")
            break

        p = r + (rsnew / rsold) * p
        rsold = rsnew

    return x

# Example to validate the solver
A = np.array([
    [4, 1, 0],
    [1, 3, -1],
    [0, -1, 2]
], dtype=float)

b = np.array([1, 2, 3], dtype=float)

# Solve the system
x = conjugate_gradient(A, b)

# Print the solution
print("Solution x:")
print(x)

# Validate the solution by computing A*x and checking against b
Ax = np.dot(A, x)
print("Computed A*x, should be close to b:")
print(Ax)

File: fortran/cgm-ai4.f90
program CGSolver
    implicit none
    integer, parameter :: dp = kind(1.0d0)
    integer, parameter :: n = 2
    double precision, dimension(n,n) :: A
    double precision, dimension(n) :: b, x
    double precision, dimension(n) :: r, p, Ap
    double precision :: rsold, rsnew, alpha, beta, tol
    integer :: i, max_iter

    ! Initialize matrix A and vector b
    A = reshape([4.0d0, 1.0d0, &
                 1.0d0, 3.0d0 ], shape=[n, n])
    b = [1.0d0, 2.0d0]
    x = [0.0d0, 0.0d0]  ! Initial guess
    tol = 1.0d-10
    max_iter = 1000

    ! Start of Conjugate Gradient method
    r = b - matmul(A, x)
    p = r
    rsold = dot_product(r, r)

    do i = 1, max_iter
        Ap = matmul(A, p)
        alpha = rsold / dot_product(p, Ap)
        x = x + alpha * p
        r = r - alpha * Ap
        rsnew = dot_product(r, r)

        if (sqrt(rsnew) < tol) then
            print *, "Convergence achieved after ", i, " iterations."
            exit
        endif

        beta = rsnew / rsold
        p = r + beta * p
        rsold = rsnew
    end do

    ! Output the solution
    print *, "Solution x:"
    print *, x

    ! Validate the solution
    print *, "A*x should be close to b:"
    print *, matmul(A, x)
end program CGSolver
